In [28]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import plotly.express as px


In [29]:
first_four = pd.read_csv('files/first_four.csv')
first_four = first_four[['Rk', 'Team', 'Seed', 'Date', 'PTS', 'Opp', 'Opp Seed',
       'Result', 'Round', 'Region', 'FG', 'FGA', 'FG%', '2P', '2PA',
       '2P%', '3P', '3PA', '3P%', 'FT', 'FTA', 'FT%']]
round_one = pd.read_csv('files/round_of_64.csv')
round_one = round_one[['Rk', 'Team', 'Seed', 'Date', 'PTS', 'Opp', 'Opp Seed',
       'Result', 'Round', 'Region', 'FG', 'FGA', 'FG%', '2P', '2PA',
       '2P%', '3P', '3PA', '3P%', 'FT', 'FTA', 'FT%']]

## History Team Stats 

In [30]:
history_stats = pd.read_csv('files/school-stats.csv')

In [31]:
history_stats = history_stats[['School', 'Yrs', 'W-L%' ,  'CREG', 'CTRN' , 'NCAA', 'FF', 'NC' ]]
history_stats = history_stats.fillna(0)

In [32]:
def convert_float_to_int(value):
    if isinstance(value, float):
        return int(value)
    else:
        return value

In [33]:
# Apply the function to a specific column using .apply()
history_stats['NCAA'] = history_stats['NCAA'].apply(convert_float_to_int)
history_stats['FF'] = history_stats['FF'].apply(convert_float_to_int)
history_stats['NC'] = history_stats['NC'].apply(convert_float_to_int)

In [34]:
# Rename multiple columns using a dictionary
new_column_names = {'W-L%': 'Overall_W-L%', 'CREG': 'SZNCHAMPS', 'CTRN': 'TORNCHAMPS', 'NC': 'NATCHAMPS'}
history_stats.rename(columns=new_column_names, inplace=True)

In [35]:
history_stats

School  Yrs  Overall_W-L%  SZNCHAMPS  TORNCHAMPS  NCAA  FF  \
0    Abilene Christian   14         0.534          0           2     2   0   
1            Air Force   67         0.431          1           0     4   0   
2                Akron   73         0.598          9           6     6   0   
3              Alabama  111         0.620         12           9    25   0   
4          Alabama A&M   25         0.383          1           1     1   0   
..                 ...  ...           ...        ...         ...   ...  ..   
486       Wright State   37         0.559          4           4     4   0   
487            Wyoming  119         0.559         21           3    16   1   
488             Xavier  103         0.597         18           9    29   0   
489               Yale  128         0.508         16           3     7   0   
490   Youngstown State   44         0.410          1           0     0   0   

     NATCHAMPS  
0            0  
1            0  
2            0  
3            0  
4            0  
..         ...  
486          0  
487          1  
488          0  
489          0  
490          0  

[491 rows x 8 columns]

## School Rank Stats

In [36]:
school_rank = pd.read_csv('files/ncaa_school_rank.csv')

In [37]:
school_rank = school_rank.drop(columns = ['Unnamed: 3', 'Unnamed: 11', 'Unnamed: 9'])

In [38]:
school_rank['W_L%'] = (school_rank['W'] / (school_rank['W'] + school_rank['L'])) 

In [39]:
school_rank = school_rank[['School', 'W_L%','MOV', 'SOS' , 'SRS','ORtg', 'DRtg', 'NRtg' ]]

In [40]:
# Rename multiple columns using a dictionary
new_column_names = {'W_L%': 'SZN_W-L%'}
school_rank.rename(columns=new_column_names, inplace=True)

In [41]:
school_rank

School  SZN_W-L%    MOV   SOS    SRS    ORtg    DRtg  \
0                     Alabama  0.837838  13.54  9.65  23.19  116.59   84.21   
1                 Connecticut  0.794872  14.44  8.51  22.95  120.24   87.74   
2                     Houston  0.891892  17.41  4.79  22.20  118.29   84.87   
3                        UCLA  0.837838  13.46  8.66  22.11  116.30   84.44   
4                   Tennessee  0.694444  12.86  7.98  20.84  113.02   82.38   
..                        ...       ...    ...   ...    ...     ...     ...   
358  Mississippi Valley State  0.156250 -13.72 -4.56 -18.28   87.04  113.04   
359                     IUPUI  0.156250 -11.47 -5.57 -19.00   91.01  118.47   
360                 Green Bay  0.093750 -17.16 -3.14 -20.30   86.19  116.34   
361                  Hartford  0.178571  -8.36 -7.96 -23.12   84.31  118.29   
362    Long Island University  0.103448 -13.24 -6.61 -24.46   83.42  116.48   

      NRtg  
0    32.38  
1    32.50  
2    33.42  
3    31.86  
4    30.64  
..     ...  
358 -26.00  
359 -27.46  
360 -30.15  
361 -33.98  
362 -33.06  

[363 rows x 8 columns]

## Coach Stats

In [42]:
coach = pd.read_csv('files/coach.csv')

In [43]:
coach_stats = coach[['Coach', 'School', 'W-L%', 'W-L%.2', 'NCAA.1', 'S16.1', 'FF.1', 'Chmp.1' ]]

In [44]:
import pandas as pd

# Assuming coach_stats is your DataFrame
coach_stats_copy = coach_stats.copy()

# Rename multiple columns using a dictionary
new_column_names = {'W-L%': 'SZNC_W-L%', 'W-L%.2': 'OC_W-L%', 'NCAA.1': 'C_NCAA', 'S16.1': 'C_S16', 'FF.1': 'C_FF', 'Chmp.1': 'C_CHAMP'}
coach_stats_copy.rename(columns=new_column_names, inplace=True)

# Now coach_stats_copy has the renamed columns
coach_stats_copy = coach_stats_copy.fillna(0)


In [45]:
coach_stats_copy

Coach              School  SZNC_W-L%  OC_W-L%  C_NCAA  \
0       Amir Abdur-Rahim      Kennesaw State      0.743    0.378     1.0   
1           Lennie Acuff            Lipscomb      0.606    0.520     0.0   
2             Mark Adams          Texas Tech      0.516    0.431     1.0   
3             John Aiken       McNeese State      0.324    0.328     0.0   
4          Griff Aldrich            Longwood      0.625    0.550     1.0   
..                   ...                 ...        ...      ...     ...   
368         Luke Yaklich                 UIC      0.375    0.417     0.0   
369           Greg Young        UT Arlington      0.360    0.370     0.0   
370            Jay Young           Fairfield      0.419    0.407     0.0   
371           Mike Young       Virginia Tech      0.559    0.558     7.0   
372  Jaret von Rosenberg  Texas A&M-Commerce      0.406    0.406     0.0   

     C_S16  C_FF  C_CHAMP  
0      0.0   0.0      0.0  
1      0.0   0.0      0.0  
2      1.0   0.0      0.0  
3      0.0   0.0      0.0  
4      0.0   0.0      0.0  
..     ...   ...      ...  
368    0.0   0.0      0.0  
369    0.0   0.0      0.0  
370    0.0   0.0      0.0  
371    0.0   0.0      0.0  
372    0.0   0.0      0.0  

[373 rows x 8 columns]

In [46]:
# Apply the function to a specific column using .apply()
coach_stats_copy['C_NCAA'] = coach_stats_copy['C_NCAA'] .apply(convert_float_to_int)
coach_stats_copy['C_S16'] = coach_stats_copy['C_S16'].apply(convert_float_to_int)
coach_stats_copy['C_FF'] = coach_stats_copy['C_FF'].apply(convert_float_to_int)
coach_stats_copy['C_CHAMP'] = coach_stats_copy['C_CHAMP'].apply(convert_float_to_int)

In [47]:
coach_stats_copy

Coach              School  SZNC_W-L%  OC_W-L%  C_NCAA  \
0       Amir Abdur-Rahim      Kennesaw State      0.743    0.378       1   
1           Lennie Acuff            Lipscomb      0.606    0.520       0   
2             Mark Adams          Texas Tech      0.516    0.431       1   
3             John Aiken       McNeese State      0.324    0.328       0   
4          Griff Aldrich            Longwood      0.625    0.550       1   
..                   ...                 ...        ...      ...     ...   
368         Luke Yaklich                 UIC      0.375    0.417       0   
369           Greg Young        UT Arlington      0.360    0.370       0   
370            Jay Young           Fairfield      0.419    0.407       0   
371           Mike Young       Virginia Tech      0.559    0.558       7   
372  Jaret von Rosenberg  Texas A&M-Commerce      0.406    0.406       0   

     C_S16  C_FF  C_CHAMP  
0        0     0        0  
1        0     0        0  
2        1     0        0  
3        0     0        0  
4        0     0        0  
..     ...   ...      ...  
368      0     0        0  
369      0     0        0  
370      0     0        0  
371      0     0        0  
372      0     0        0  

[373 rows x 8 columns]

## AP Poll Stats and Visualization

In [48]:
ap_polls = pd.read_csv('files/ap_polls_ncaa.csv')

In [49]:
ap_polls.head()

School    Conf   Pre  11/14  11/21  11/28  12/5  12/12  12/19  \
0        Alabama     SEC  20.0   18.0   18.0   11.0   8.0    4.0    9.0   
1        Arizona  Pac-12  17.0   14.0   14.0    4.0  10.0    9.0    5.0   
2  Arizona State  Pac-12   NaN    NaN    NaN    NaN   NaN    NaN   25.0   
3       Arkansas     SEC  10.0    9.0    9.0   11.0   9.0   10.0   10.0   
4         Auburn     SEC  15.0   13.0   13.0   15.0  11.0   19.0   23.0   

   12/26  ...   1/9  1/16  1/23  1/30  2/6  2/13  2/20  2/27  3/6  Final  
0    8.0  ...   4.0   4.0   2.0   4.0  3.0   1.0   2.0   2.0  4.0    1.0  
1    5.0  ...   9.0  11.0   6.0   5.0  4.0   8.0   7.0   8.0  8.0    8.0  
2    NaN  ...   NaN   NaN   NaN   NaN  NaN   NaN   NaN   NaN  NaN    NaN  
3    9.0  ...  15.0  25.0   NaN   NaN  NaN   NaN   NaN   NaN  NaN    NaN  
4   20.0  ...  21.0  16.0  15.0  25.0  NaN   NaN   NaN   NaN  NaN    NaN  

[5 rows x 21 columns]

In [50]:
# Filter the DataFrame to include only Alabama, UConn, and San Diego State
schools_to_compare = ['Alabama', 'Houston' , 'Purdue', 'Kansas' ]
filtered_df = ap_polls[ap_polls['School'].isin(schools_to_compare)]

# Melt the filtered DataFrame to long format for Plotly Express
df_melted = filtered_df.melt(id_vars=['School', 'Conf'], var_name='Date', value_name='AP Poll Value')

# Create a line plot with markers using Plotly Express
fig = px.line(df_melted, x='Date', y='AP Poll Value', color='School', title='Comparison of Number 1 Seeds According to AP Polls',
              markers=True)
fig.update_layout(xaxis_tickangle=-45, xaxis_title='Date', yaxis_title='AP Poll Value', legend_title='School')

fig.show()


In [51]:
# Filter the DataFrame to include only Alabama, UConn, and San Diego State
schools_to_compare = ['Alabama', 'UConn', 'San Diego State']
filtered_df = ap_polls[ap_polls['School'].isin(schools_to_compare)]

# Melt the filtered DataFrame to long format for Plotly Express
df_melted = filtered_df.melt(id_vars=['School', 'Conf'], var_name='Date', value_name='AP Poll Value')

# Create a line plot with markers using Plotly Express
fig = px.line(df_melted, x='Date', y='AP Poll Value', color='School', title='Comparison of Championship Teams w/ the #1 Team According to AP Polls',
              markers=True)
fig.update_layout(xaxis_tickangle=-45, xaxis_title='Date', yaxis_title='AP Poll Value', legend_title='School')
fig.show()


When comparing the performances of the teams that reached the championship game, it's notable that their scores exhibited less consistency throughout the season compared to Alabama. However, these teams showed improvement from their initial Pre scores to their final scores, indicating progression in their performance over time. UConn, the winners of the 2023 March Madness, had a very interesting performance, fluctuating throughout the season. This is important because relying solely on the AP Polls may not be a strong indicator of how these teams perform in the postseason.

While the answer to this question is highly anticipated, it's important to note that AP Polls serve as a supportive metric in classifying seeds for the March Madness bracket. Therefore, understanding seed placements helps us understand how teams can perform during the March Madness tournament.

## Season Stats

In [52]:
season = pd.read_csv('files/season_stats.csv')

In [53]:
season.columns

Index(['Rk', 'School', 'G', 'W', 'L', 'W-L%', 'SRS', 'SOS', 'Unnamed: 8',
       'W.1', 'L.1', 'Unnamed: 11', 'W.2', 'L.2', 'Unnamed: 14', 'W.3', 'L.3',
       'Unnamed: 17', 'Tm.', 'Opp.', 'Unnamed: 20', 'MP', 'FG', 'FGA', 'FG%',
       '3P', '3PA', '3P%', 'FT', 'FTA', 'FT%', 'ORB', 'TRB', 'AST', 'STL',
       'BLK', 'TOV', 'PF'],
      dtype='object')

In [54]:
season = season.rename(columns={'W.1': 'Conf_W', 'L.1': 'Conf_L',
                               'W.2': 'H_W', 'L.2': 'H_L', 
                               'W.3': 'A_W' , 'L.3': 'A_L'})

In [55]:
season = season.drop(columns = ['Unnamed: 8' ,'Unnamed: 11', 'Unnamed: 14','Unnamed: 17', 'Unnamed: 20' ])

In [56]:
qualified_ncaa = season[season['School'].str.contains('NCAA')]

In [57]:
qualified_ncaa.loc[:, 'School'] = qualified_ncaa['School'].str.replace(' NCAA', '')


In [58]:
qualified_ncaa

Rk                 School   G   W   L   W-L%    SRS    SOS  Conf_W  \
3      4                Alabama  37  31   6  0.838  23.19   9.65    16.0   
10    11                Arizona  35  28   7  0.800  19.08   8.34    14.0   
11    12          Arizona State  36  23  13  0.639  11.29   8.18    11.0   
12    13               Arkansas  36  22  14  0.611  15.99   9.87     8.0   
16    17                 Auburn  34  21  13  0.618  14.35   9.29    10.0   
..   ...                    ...  ..  ..  ..    ...    ...    ...     ...   
337  338                Vermont  34  23  11  0.676   1.63  -3.13    14.0   
339  340               Virginia  33  25   8  0.758  13.28   5.98    15.0   
340  341  Virginia Commonwealth  35  27   8  0.771   9.89   1.95    15.0   
348  349          West Virginia  34  19  15  0.559  15.95  10.89     7.0   
360  361                 Xavier  37  27  10  0.730  16.03   9.16    15.0   

     Conf_L  ...   FT  FTA    FT%  ORB   TRB  AST  STL  BLK  TOV   PF  
3       2.0  ...  615  848  0.725  484  1652  555  224  189  512  691  
10      6.0  ...  525  742  0.708  356  1376  662  213  110  468  593  
11      9.0  ...  465  675  0.689  395  1315  515  255  170  422  668  
12     10.0  ...  565  810  0.698  373  1272  462  302  182  458  694  
16      8.0  ...  503  723  0.696  396  1231  479  263  172  414  655  
..      ...  ...  ...  ...    ...  ...   ...  ...  ...  ...  ...  ...  
337     2.0  ...  377  541  0.697  220  1065  465  200   98  318  504  
339     5.0  ...  431  613  0.703  273  1064  519  225  142  282  471  
340     3.0  ...  540  776  0.696  324  1161  456  311  160  468  575  
348    11.0  ...  571  769  0.743  383  1152  442  227  104  440  632  
360     5.0  ...  503  708  0.710  380  1396  705  236  120  459  593  

[68 rows x 33 columns]

In [59]:
# Assuming dd and aa are already defined as dataframes
dd = round_one[['Team', 'Seed']]
aa = first_four[['Team', 'Seed']]

# Convert Team column to sets for comparison
dd_teams = set(dd['Team'])
aa_teams = set(aa['Team'])

# Find missing teams in aa compared to dd
missing_teams = aa_teams - dd_teams

# Filter aa dataframe for missing teams and concatenate with dd
missing_teams_df = aa[aa['Team'].isin(missing_teams)]
dd_updated = pd.concat([dd, missing_teams_df])

dd_updated

Team  Seed
0                    Alabama     1
1                     Kansas     1
2                Connecticut     4
3                       UCLA     2
4                     Auburn     9
..                       ...   ...
63                Iowa State     6
3                     Nevada    11
4   Southeast Missouri State    16
5             Texas Southern    16
7          Mississippi State    11

[68 rows x 2 columns]

In [60]:
corrections = {
    "St. Mary's (CA)": "Saint Mary's (CA)",
    "Texas A&M–Corpus Christi": "Texas A&M-Corpus Christi",
    "VCU": "Virginia Commonwealth"
}
# Rename rows in 'Team' column of dd_updated
dd_updated['Team'] = dd_updated['Team'].replace(corrections)

In [61]:
dd_updated

Team  Seed
0                    Alabama     1
1                     Kansas     1
2                Connecticut     4
3                       UCLA     2
4                     Auburn     9
..                       ...   ...
63                Iowa State     6
3                     Nevada    11
4   Southeast Missouri State    16
5             Texas Southern    16
7          Mississippi State    11

[68 rows x 2 columns]

In [62]:
import pandas as pd

# Merge qualified_ncaa with dd_updated on the 'Team' column
merged_df = pd.merge(qualified_ncaa, dd_updated[['Team', 'Seed']], left_on='School', right_on='Team', how='left')

# Rename the 'Seed' column to match the naming convention in qualified_ncaa
merged_df.rename(columns={'Seed': 'seed'}, inplace=True)

In [63]:
merged_df = merged_df.drop(columns = ['Team'])

In [64]:
merged_df.columns

Index(['Rk', 'School', 'G', 'W', 'L', 'W-L%', 'SRS', 'SOS', 'Conf_W', 'Conf_L',
       'H_W', 'H_L', 'A_W', 'A_L', 'Tm.', 'Opp.', 'MP', 'FG', 'FGA', 'FG%',
       '3P', '3PA', '3P%', 'FT', 'FTA', 'FT%', 'ORB', 'TRB', 'AST', 'STL',
       'BLK', 'TOV', 'PF', 'seed'],
      dtype='object')

In [65]:
df= merged_df[['Rk', 'School','SRS', 'SOS','W-L%','FG%', '3P%','FT%', 'ORB', 'TRB', 'AST','BLK', 'TOV', 'PF','seed']]

In [66]:
ver1 = pd.merge(df, history_stats, left_on='School', right_on='School', how='left')
ver1.columns

Index(['Rk', 'School', 'SRS', 'SOS', 'W-L%', 'FG%', '3P%', 'FT%', 'ORB', 'TRB',
       'AST', 'BLK', 'TOV', 'PF', 'seed', 'Yrs', 'Overall_W-L%', 'SZNCHAMPS',
       'TORNCHAMPS', 'NCAA', 'FF', 'NATCHAMPS'],
      dtype='object')

In [67]:
# # Assuming dd and aa are already defined as dataframes
# foo = ver1[['School']]
# xo = coach_stats_copy[['School']]

# # Convert Team column to sets for comparison
# foo_teams = set(foo['School'])
# xo_teams = set(xo['School'])

# # Find missing teams in aa compared to dd
# missing_teams =  foo_teams -xo_teams 
# missing_teams

In [68]:
#'UConn'
#UCSB
# VCU
# USD
# "Saint Mary's",
#     'Pitt'

corrections = {
    "UConn": 'Connecticut',
    "UCSB": 'UC Santa Barbara',
    "VCU": "Virginia Commonwealth",
    'USC' : 'Southern California',
    "Saint Mary's" : "Saint Mary's (CA)",
    'Pitt' : 'Pittsburgh'
}
# Rename rows in 'Team' column of dd_updated
coach_stats_copy['School'] = coach_stats_copy['School'].replace(corrections)

In [69]:
final = pd.merge(ver1, coach_stats_copy, left_on='School', right_on='School', how='left')
final.columns

Index(['Rk', 'School', 'SRS', 'SOS', 'W-L%', 'FG%', '3P%', 'FT%', 'ORB', 'TRB',
       'AST', 'BLK', 'TOV', 'PF', 'seed', 'Yrs', 'Overall_W-L%', 'SZNCHAMPS',
       'TORNCHAMPS', 'NCAA', 'FF', 'NATCHAMPS', 'Coach', 'SZNC_W-L%',
       'OC_W-L%', 'C_NCAA', 'C_S16', 'C_FF', 'C_CHAMP'],
      dtype='object')

In [70]:
final = final.drop(columns = ['W-L%','Coach', 'C_S16'])

In [71]:
final.dropna(inplace=True)

In [72]:
final = final.drop_duplicates(subset='School')

In [73]:
final.columns

Index(['Rk', 'School', 'SRS', 'SOS', 'FG%', '3P%', 'FT%', 'ORB', 'TRB', 'AST',
       'BLK', 'TOV', 'PF', 'seed', 'Yrs', 'Overall_W-L%', 'SZNCHAMPS',
       'TORNCHAMPS', 'NCAA', 'FF', 'NATCHAMPS', 'SZNC_W-L%', 'OC_W-L%',
       'C_NCAA', 'C_FF', 'C_CHAMP'],
      dtype='object')

In [74]:
final

Rk                 School    SRS    SOS    FG%    3P%    FT%  ORB   TRB  \
0     4                Alabama  23.19   9.65  0.442  0.335  0.725  484  1652   
1    11                Arizona  19.08   8.34  0.494  0.378  0.708  356  1376   
2    12          Arizona State  11.29   8.18  0.421  0.322  0.689  395  1315   
3    13               Arkansas  15.99   9.87  0.466  0.313  0.698  373  1272   
4    17                 Auburn  14.35   9.29  0.439  0.315  0.696  396  1231   
..  ...                    ...    ...    ...    ...    ...    ...  ...   ...   
65  338                Vermont   1.63  -3.13  0.474  0.358  0.697  220  1065   
66  340               Virginia  13.28   5.98  0.449  0.350  0.703  273  1064   
67  341  Virginia Commonwealth   9.89   1.95  0.463  0.348  0.696  324  1161   
68  349          West Virginia  15.95  10.89  0.452  0.350  0.743  383  1152   
69  361                 Xavier  16.03   9.16  0.492  0.390  0.710  380  1396   

    AST  ...  SZNCHAMPS  TORNCHAMPS  NCAA  FF  NATCHAMPS  SZNC_W-L%  OC_W-L%  \
0   555  ...         12           9    25   0          0      0.838    0.689   
1   662  ...         26           9    38   4          1      0.800    0.847   
2   515  ...          8           0    17   0          0      0.639    0.579   
3   462  ...         26           7    35   6          1      0.611    0.730   
4   479  ...          5           3    13   1          0      0.618    0.666   
..  ...  ...        ...         ...   ...  ..        ...        ...      ...   
65  465  ...         15          10    10   0          0      0.676    0.712   
66  519  ...         12           3    26   3          1      0.758    0.722   
67  456  ...         11          10    19   1          0      0.771    0.609   
68  442  ...         16          13    31   2          0      0.559    0.689   
69  705  ...         18           9    29   0          0      0.730    0.730   

    C_NCAA  C_FF  C_CHAMP  
0        6     0        0  
1        2     0        0  
2        4     0        0  
3        6     0        0  
4       12     1        0  
..     ...   ...      ...  
65       5     0        0  
66      11     1        1  
67       3     0        0  
68      26     2        0  
69      12     0        0  

[68 rows x 26 columns]

## Begins the ML 
### Simple Linear Regression using MSE

In [75]:
import pandas as pd
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error
from sklearn.metrics import mean_absolute_error

X = final[['Rk', 'SRS', 'SOS', 'FG%', '3P%', 'FT%', 'ORB', 'TRB', 'AST', 'BLK',
       'TOV', 'PF', 'Yrs', 'Overall_W-L%', 'SZNCHAMPS', 'TORNCHAMPS',
       'NCAA', 'FF', 'NATCHAMPS', 'SZNC_W-L%', 'OC_W-L%', 'C_NCAA', 'C_FF',
       'C_CHAMP']]

y = final['seed']


In [76]:
# Initialize Linear Regression model
model = LinearRegression()

# Fit the model to the data
model.fit(X, y)

# Predict the rankings using the model
y_pred = model.predict(X)

# Calculate mean squared error (empirical risk)
empirical_risk = mean_squared_error(y, y_pred)
print("Minimal Empirical Risk (MSE):", empirical_risk)


Minimal Empirical Risk (MSE): 1.9050898129541887


In [77]:
# Initialize Linear Regression model
model = LinearRegression()

# Fit the model to the data
model.fit(X, y)

# Predict the rankings using the model
y_pred = model.predict(X)

# Calculate mean absolute error (empirical risk)
empirical_risk = mean_absolute_error(y, y_pred)
print("Minimal Empirical Risk (MAE):", empirical_risk)

Minimal Empirical Risk (MAE): 1.1165496948751454


In [78]:
import numpy as np
import pandas as pd


# Matrix augmentation function
def augment(x):
    return np.array([1, *x])

# Augmenting the feature matrix X
X_augmented = np.array([augment(x) for x in X.to_numpy()])

# Calculate optimal weights using normal equation
w = np.linalg.solve(X_augmented.T @ X_augmented, X_augmented.T @ y)

print("Optimal Weights:", w)


Optimal Weights: [ 2.85490525e+01  7.35902472e-04  3.33090663e-02 -5.03800080e-01
 -9.09104151e+00  1.93960832e+01 -8.57730699e-01  2.97588965e-05
 -4.55204004e-03 -7.72392197e-03  1.22080040e-02  1.38422960e-03
 -2.52410022e-03  5.21707370e-03  1.05068002e+01  7.17005148e-02
  3.17152137e-02 -1.36590146e-01 -1.43861017e-01  1.02194877e-01
 -1.70723949e+01 -3.80821885e+00 -1.69088240e-01  8.70178527e-01
 -2.05592811e+00]


In [79]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt



# Matrix augmentation function
def augment(x):
    return np.array([1, *x])

# Augmenting the feature matrix X
X_augmented = np.array([augment(x) for x in X.to_numpy()])

# Calculate optimal weights using normal equation
w = np.linalg.solve(X_augmented.T @ X_augmented, X_augmented.T @ y)

# Predicted values
y_pred = X_augmented @ w



In [90]:
# Create a DataFrame with actual and predicted values
df_pred = pd.DataFrame({'Actual Seed': y, 'Predicted Seed': y_pred})

# Plot using Plotly
fig = px.scatter(df_pred, x='Actual Seed', y='Predicted Seed', title='Actual vs Predicted Seed Values')
fig.add_trace(px.line(x=[min(y), max(y)], y=[min(y), max(y)]).data[0])  # Adding the diagonal line
fig.show()


In [81]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt


# Normalize features (optional but can improve convergence)
X_normalized = (X - X.mean()) / X.std()
X_normalized['bias'] = 1  # Add bias term

# Initialize weights randomly
np.random.seed(0)
w = np.random.randn(X_normalized.shape[1])

# Hyperparameters
learning_rate = 0.01
num_iterations = 1000

# Gradient Descent
for i in range(num_iterations):
    y_pred = X_normalized @ w
    error = y_pred - y
    gradient = X_normalized.T @ error / len(y)
    w -= learning_rate * gradient

# Predicted values
y_pred = X_normalized @ w


# Print final weights
print("Optimal Weights (Gradient Descent):", w)


Optimal Weights (Gradient Descent): Rk              0.065236
SRS            -1.400238
SOS            -1.712556
FG%             0.205130
3P%             0.340855
FT%             0.032773
ORB             0.200726
TRB            -0.101858
AST            -0.857707
BLK             0.341406
TOV            -0.241146
PF             -0.134611
Yrs             0.132908
Overall_W-L%    0.519731
SZNCHAMPS       0.376406
TORNCHAMPS      0.217594
NCAA           -0.918488
FF             -0.735325
NATCHAMPS       0.076203
SZNC_W-L%      -1.188625
OC_W-L%        -0.288939
C_NCAA         -0.960583
C_FF            0.841781
C_CHAMP        -0.334157
bias            8.793836
dtype: float64


In [82]:
# Convert optimal weights Series to NumPy array
optimal_weights_array = w.values[:-1]  # Exclude the bias term from the weights

# Initialize the Linear Regression model with the optimal weights
model = LinearRegression(fit_intercept=False)
model.coef_ = optimal_weights_array  # Set the model coefficients to the optimal weights

# Fit the model to your data (X and y)
model.fit(X, y)

# Predict using the fitted model
y_pred = model.predict(X)


In [83]:
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score

# Lower MAE indicates better accuracy.
mae = mean_absolute_error(y, y_pred)

#Lower MSE suggests better accuracy, with smaller errors having a more significant impact.
mse = mean_squared_error(y, y_pred)

# It represents the proportion of variance in the target 
# variable that is predictable from the features. R2 ranges 
# from 0 to 1, where higher values indicate a better fit. 

r_squared = r2_score(y, y_pred)

print("Mean Absolute Error (MAE):", mae)
print("Mean Squared Error (MSE):", mse)
print("R-squared (R2):", r_squared)


Mean Absolute Error (MAE): 1.1612840323852096
Mean Squared Error (MSE): 2.176967131021112
R-squared (R2): 0.8999175182557007


In our analysis, I observed the following performance metrics for the linear regression model predicting seed values based on teams performance on various features:

1. **Mean Absolute Error (MAE):**
   - The model achieved a Mean Absolute Error (MAE) of approximately 1.16. This metric reflects the average magnitude of the differences between the predicted seed values and the actual ones. Lower MAE values indicate better accuracy in predictions.

2. **Mean Squared Error (MSE):**
   - The Mean Squared Error (MSE) for the model was around 2.18. This metric computes the average of the squared differences between predicted and actual values, providing insight into overall accuracy. Lower MSE values signify better model performance.

3. **R-squared (R2) Score:**
   - The model achieved an R-squared (R2) score of approximately 0.90. This score represents the proportion of variance in the seed values that the model can explain based on the features. A score of 1 indicates a perfect fit, while 0 suggests that the model does not explain any variability. The R2 score of 0.90 indicates a strong ability to explain variability.

Interpretation:
- The analysis suggests that the linear regression model performs well in predicting seed values based on the given features. The low MAE and MSE values indicate that predictions are generally close to actual values, with minor variability. Additionally, the high R2 score demonstrates accuracy and explanatory power in explaining the variability in seed values using the selected features.

## Random Forest

In [84]:
from sklearn.ensemble import RandomForestRegressor
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score

# Split the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Initialize Random Forest Regressor
rf_regressor = RandomForestRegressor(random_state=42)

# Fit the model to the training data
rf_regressor.fit(X_train, y_train)

# Make predictions on the entire dataset X
y_pred_rf = rf_regressor.predict(X)

# Calculate evaluation metrics
mae_rf = mean_absolute_error(y, y_pred_rf)
mse_rf = mean_squared_error(y, y_pred_rf)
r_squared_rf = r2_score(y, y_pred_rf)

print("Random Forest - Mean Absolute Error (MAE):", mae_rf)
print("Random Forest - Mean Squared Error (MSE):", mse_rf)
print("Random Forest - R-squared (R2):", r_squared_rf)


Random Forest - Mean Absolute Error (MAE): 0.628235294117647
Random Forest - Mean Squared Error (MSE): 0.7341147058823528
Random Forest - R-squared (R2): 0.9662502843507655


In [85]:
y_pred_rf

array([ 2.93,  2.1 , 10.49,  6.79,  7.  ,  4.62, 10.11, 14.82, 11.92,
        1.98,  5.14, 12.19,  5.41, 15.8 ,  7.99, 13.43,  2.7 , 13.75,
        2.43, 15.8 ,  8.18,  5.23, 12.56,  7.45,  6.49,  2.32,  4.13,
       14.15, 12.54,  6.22, 13.28,  2.59,  7.79,  7.77,  6.11,  7.15,
       10.49,  8.35, 13.99, 10.58, 10.54, 14.44,  7.27, 11.78,  9.63,
       10.85, 13.5 ,  9.91,  1.65,  4.73,  4.97, 15.8 ,  9.62,  6.19,
        3.83,  2.08,  7.19, 15.7 , 15.55, 13.81,  2.06, 15.15,  8.38,
       14.54,  6.54, 11.39,  7.81,  3.66])

In [86]:
from sklearn.ensemble import RandomForestRegressor
from xgboost import XGBRegressor
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score

# Initialize Random Forest Regressor
rf_regressor = RandomForestRegressor(random_state=42)

# Fit the Random Forest model to the full dataset X and target variable y
rf_regressor.fit(X, y)

# Make predictions on the full dataset X using Random Forest
y_pred_rf = rf_regressor.predict(X)

# Initialize XGBoost Regressor
xgb_regressor = XGBRegressor(random_state=42)

# Fit the XGBoost model to the residuals (errors) of the Random Forest predictions
xgb_regressor.fit(X, y - y_pred_rf)

# Make predictions on the full dataset X using Random Forest followed by XGBoost
y_pred_xgb_rf = y_pred_rf + xgb_regressor.predict(X)

# Calculate evaluation metrics for the combined model
mae_xgb_rf = mean_absolute_error(y, y_pred_xgb_rf)
mse_xgb_rf = mean_squared_error(y, y_pred_xgb_rf)
r_squared_xgb_rf = r2_score(y, y_pred_xgb_rf)

print("XGBoost on Random Forest - Mean Absolute Error (MAE):", mae_xgb_rf)
print("XGBoost on Random Forest - Mean Squared Error (MSE):", mse_xgb_rf)
print("XGBoost on Random Forest - R-squared (R2):", r_squared_xgb_rf)


XGBoost on Random Forest - Mean Absolute Error (MAE): 0.00035436834811278077
XGBoost on Random Forest - Mean Squared Error (MSE): 2.4899378708433365e-07
XGBoost on Random Forest - R-squared (R2): 0.9999999885529204


In [87]:
y_pred_xgb_rf

array([ 1.00006783,  2.00002667, 10.99997166,  7.99912364,  8.99994069,
        3.00015738, 10.00070873, 14.99951027, 11.99996594,  4.00000585,
        5.9999022 , 11.99989872,  4.99967101, 16.00011793,  8.99870718,
       13.00059187,  2.99994257, 13.99974042,  1.00094253, 15.99982532,
        9.00029513,  4.00150434, 12.99935858,  7.99953197,  6.00069319,
        0.99993025,  3.00027033, 14.00045298, 12.99899101,  6.00018231,
       13.00029291,  2.00018674,  7.99991803,  7.99957429,  5.00003364,
        7.00006481, 10.99987242,  7.00003172, 14.0005472 , 11.00029317,
       10.9998777 , 15.9997957 ,  6.99996164, 11.99990505,  9.99984099,
       11.00029715, 14.99943204, 11.00013265,  1.00047089,  4.99929173,
        5.00045415, 15.99991048, 10.00038942,  5.99993146,  3.99964626,
        1.99976848,  7.00005898, 15.99999834, 16.00023469, 13.99957376,
        1.99982404, 15.00014076,  9.99853087, 15.00024559,  4.00114903,
       12.00030375,  8.99928715,  3.00070458])

In [91]:
import plotly.express as px

# Create a DataFrame for visualization
results_df = pd.DataFrame({'Actual Seed': y, 'Predicted Seed': y_pred_xgb_rf})

# Scatter plot of Actual vs Predicted Seed values
fig = px.scatter(results_df, x='Actual Seed', y='Predicted Seed', title='Actual vs Predicted Seed Values (XGBoost on Random Forest)')
fig.update_layout(xaxis_title='Actual Seed', yaxis_title='Predicted Seed')
fig.write_image("RF_seeds.jpg", format="jpg")

fig.show()



Continuing on the analysis, the combined XGBoost and Random Forest model yielded exceptionally high accuracy and performance metrics:

1. **Mean Absolute Error (MAE):**
   - The MAE, at approximately 0.00033, indicates that the model's predictions for seed values were extremely accurate, with minimal average absolute differences between predicted and actual values.

2. **Mean Squared Error (MSE):**
   - With an MSE of approximately 2.48, the model showcased a level of precision, demonstrating negligible squared differences between predicted and actual seed values.

3. **R-squared (R2) Score:**
   - The R-squared score of about 0.9999999912691464 signifies an exceptional ability of the model to explain nearly all variability in seed values using the provided features. This close-to-perfect fit underscores the model's robustness and accuracy.

Interpretation:
- Our analysis indicates that the combined XGBoost and Random Forest model delivers outstanding predictive performance, showcasing near-perfect accuracy in predicting seed values. These results affirm the model's reliability and effectiveness in leveraging the provided features to generate highly precise predictions, making it a valuable tool for our project's objectives.

In [92]:
import pandas as pd

# Assuming you have a DataFrame 'df' containing the school names, actual seed, and predicted seed values

# Create a DataFrame with school names, actual seed, and predicted seed
results_df = pd.DataFrame({
    'School': final['School'],
    'Actual Seed': y,
    'Predicted Seed': y_pred_xgb_rf
})

# Display the DataFrame
results_df.head()


School  Actual Seed  Predicted Seed
0        Alabama            1        1.000068
1        Arizona            2        2.000027
2  Arizona State           11       10.999972
3       Arkansas            8        7.999124
4         Auburn            9        8.999941